R/latent_individual.R

```diff
@@ -108,28 +108,77 @@ is_latent_individual <- function(data) {
   inherits(data, "epidist_latent_individual")
 }
 
+#' Check if data has the `epidist_latent_individual` class
+#'
+#' @param data A `data.frame` or `data.table` containing line list data
+#' @param family Output of a call to `brms::brmsfamily()`
+#'
+#' @importFrom rstan lookup
+#' @method epidist_family epidist_latent_individual
+#' @family latent_individual
+#' @export
+epidist_family.epidist_latent_individual <- function(data,
+                                                     family = brms::lognormal(),
+                                                     ...) {
+  epidist_validate(data)
+  checkmate::assert_class(family, "brmsfamily")
+  # brms also has lb and ub in brms::custom_family, but not in brms::brmsfamily
+  non_mu_links <- family[[paste0("link_", setdiff(family$dpars, "mu"))]]
+  brms::custom_family(
+    paste0("latent_", family$family),
+    dpars = family$dpars,
+    links =   c(family$link, non_mu_links),
+    type = family$type,
+    vars = c("pwindow", "swindow", "vreal1"),
+    loop = FALSE
+  )
+}
+
 #' Define a formula for the latent_individual model
 #'
 #' @param data ...
-#' @param delay_central Formula for the delay mean. Defaults to intercept only.
-#' @param sigma Formula for the delay standard deviation. Defaults to intercept
-#' only.
+#' @param family The output of `epidist_family()`
+#' @param formula A list of formula
 #' @param ... ...
 #' @method epidist_formula epidist_latent_individual
 #' @family latent_individual
 #' @importFrom stats as.formula
 #' @export
-epidist_formula.epidist_latent_individual <- function(data, delay_central = ~ 1,
-                                                      sigma = ~ 1, ...) {
+epidist_formula.epidist_latent_individual <- function(data, family, formula,
+                                                      ...) {
   epidist_validate(data)
-  if (!inherits(delay_central, "formula")) {
-    cli::cli_abort("A valid formula for delay_central must be provided")
+  if (!all(sapply(formula, inherits, "formula"))) {
+    cli::cli_abort("formula must contain a list of formula")
   }
-  if (!inherits(sigma, "formula")) {
-    cli::cli_abort("A valid formula for sigma must be provided")
+  required_dpars <- family$dpars
```

Comment:
Is there anything in here that is model specific?

Reply:
No the `family$dpars` are not model specific they're family specific (where family here means `brmsfamily` rather than `custom_family` -- which has model specific parts in it).

So extrapolating this, it's unclear to me the extent to which we need this function to dispatch on the model class. But seeing as we just have one model I think it's fine to leave it as it is for now and consider how we extend to a second model at a later point.